core/multi_root_tree/multi_root_dag_tree.go

```diff
@@ -0,0 +1,101 @@
+package multi_root_tree
+
+import (
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrCyclicDependencyEncountered is triggered a tree has a cyclic dependency
+	ErrCyclicDependencyEncountered = errors.New("a cycle dependency encountered in the tree")
+)
+
+// MultiRootDAGTree - represents a data type which has multiple independent root nodes
+// all root nodes have their independent tree based on depdencies of TreeNode.
+// it also maintains a map of nodes for faster lookups and managing node data.
+type MultiRootDAGTree struct {
```

Comment:
Struct name and package name are prefixed with the same literal and will stutter when used. Can we rename it to something else?

Reply:
we can call this package as tree

Reply:
updated